### PR TITLE
Build and test TVM under minimal configuration

### LINKED ISSUE #11932: Exercise TVM under minimal configuration in CI
We have seen a couple bugs due to microTVM being presumed-ON in config.cmake. Namely, you get python errors importing TVM right now when USE_MICRO is OFF. We should have a regression test that verifies basic functionality with everything (or nearly everything) OFF.

Context: apache/tvm#9617
And another micro-related issue of the same kind, which i don't have handy right now.

cc @gigiblender

### PR DESCRIPTION
**Pull request by gigiblender**
This PR builds and tests TVM (running the CPP and unittests) under minimal configuration with some debug flags enabled:
- `USE_RELAY_DEBUG=ON` in TVM
- `-Wp,-D_GLIBCXX_ASSERTIONS` in TVM
- `-DLLVM_ENABLE_ASSERTIONS=ON` in LLVM

It also adds this configuration to the CI. 

`tests/python/unittest/test_meta_schedule_task_scheduler.py::test_meta_schedule_task_scheduler_multiple_gradient_based` results in an array OOB access and a segfault due to `D_GLIBCXX_ASSERTIONS`. I disable this test for now and will open an issue to solve it ASAP.

It should fix #11932 and address [this discussion](https://discuss.tvm.apache.org/t/pre-rfc-new-ci-container-ci-cpu-asserts/12536/9).

### CODE REVIEW

File: ci/jenkins/Test.groovy.j2
Comment (driazati) [inline ci/jenkins/Test.groovy.j2:237]: can you use the `m.test_step` macro here instead? That way we don't need to duplicate all the meta-logic (timeout, workspace, etc) just for this test

File: ci/jenkins/Test.groovy.j2
Comment (driazati) [inline ci/jenkins/Test.groovy.j2:199]: can you also use the `m.download_artifacts` macro here?

File: tests/python/unittest/test_meta_schedule_task_scheduler.py
Comment (driazati) [inline tests/python/unittest/test_meta_schedule_task_scheduler.py:340]: Can you file and link an issue here with the context? Also it'd be nice if this was a `skipIf`, maybe you can use a combo of flags from libinfo https://github.com/apache/tvm/blob/main/src/support/libinfo.cc#L238?

File: tests/scripts/task_config_build_minimal.sh
Comment (driazati) [inline tests/scripts/task_config_build_minimal.sh:26]: We've had some issues come up before where building in debug mode surfaced errors that weren't there in release mode, this is somewhat of a different discussion but I wonder if we should also set `CMAKE_BUILD_TYPE=Debug` here to check that in CI (I guess the question is whether that should be its own test or rolled into this one)

**Review by driazati — COMMENTED** [review]
Thanks for tackling this! It's been on the issue list for quiet a while, things mostly lgtm. Can you also follow up to add an ECR repo for `ci_minimal` as well? https://github.com/tlc-pack/ci/blob/main/terraform/vars/tvm-ci-prod.auto.tfvars#L79

**Comment by areusch** [comment]
cc @kparzysz-quic @Mousius @manupa-arm @cbalint13 @tqchen

File: tests/scripts/task_cpp_unittest.sh
Comment (areusch) [inline tests/scripts/task_cpp_unittest.sh:49]: i really wanted to be able to have you just look this up in CMakeCache.txt, but unfortunately we don't cache these. i wonder if we should be setting this with `set(USE_ABC ___ CACHED)` so we can inspect that. what if we modified [Summary.cmake](https://github.com/apache/tvm/blob/main/cmake/utils/Summary.cmake) to write the options to a file via https://cmake.org/cmake/help/latest/command/file.html#append?

Comment (driazati) [reply tests/scripts/task_cpp_unittest.sh:49]: it might also be cleaner to split this out into `task_cpp_unittest_micro.sh` or something and just add it as another step everywhere that currently runs the cpp unit tests, that way its easy to run regardless of the specific build options

File: ci/jenkins/Test.groovy.j2
Comment (gigiblender) [reply ci/jenkins/Test.groovy.j2:237]: I'm not sure how to do that. The `test_step` inlines the body while in this case, I am calling `run_unittest_minimal`. I could try to inline the method and use the macro, but that might not work due to the 64KB bytecode size/method limit in the JVM.

I also guess I could modify the macros.

Comment (gigiblender) [reply ci/jenkins/Test.groovy.j2:237]: Ok, I am adding a new macro

Comment (driazati) [reply ci/jenkins/Test.groovy.j2:237]: I meant something like https://github.com/apache/tvm/blob/main/ci/jenkins/Test.groovy.j2#L202-L218, if you do that it will automatically call the test method

Comment (gigiblender) [reply ci/jenkins/Test.groovy.j2:220]: Yes, but I needed a different macro that did not include the step name (eg `unittest: CPU MINIMAL`, `frontend: aarch64 2 of 2`, `unittest: CPU`). 

https://github.com/apache/tvm/pull/12178/files#diff-1a31fe4681520821e63f0bec0a963c86b8eea7af2c7130b28eba0b987a5fc21bR86

File: tests/scripts/task_config_build_minimal.sh
Comment (gigiblender) [reply tests/scripts/task_config_build_minimal.sh:26]: Enabled debug mode.

File: tests/scripts/task_cpp_unittest.sh
Comment (gigiblender) [reply tests/scripts/task_cpp_unittest.sh]: I went the route of separating the script into two

File: ci/jenkins/Test.groovy.j2
Comment (driazati) [inline ci/jenkins/Test.groovy.j2:222]: I still don't understand why we need a new macro, does this change not do the same thing with existing code?
```suggestion
  {% call m.test_step(
    name="unittest: CPU MINIMAL",
    node="CPU-SMALL",
    ws="tvm/ut-python-cpu-minimal",
    platform="minimal",
    docker_image="ci_minimal",
  ) %}
    {{ m.download_artifacts(tag='cpu-minimal', filenames=tvm_lib) }}
    cpp_unittest(ci_minimal)
    python_unittest(ci_minimal)
  {% endcall %}
```

File: tests/scripts/ci.py
Comment (driazati) [inline tests/scripts/ci.py:622]: iirc these turn into command line flags

```suggestion
            "micro-cpp": CPP_MICRO_UNITTEST,
```

File: ci/jenkins/Test.groovy.j2
Comment (gigiblender) [reply ci/jenkins/Test.groovy.j2:222]: Not really. The `test_step` macro will expand and generate something as if the body of run_unittest_minimal was inlined at the call site. 

Due to the JVM method size limit, I cannot inline the body of `run_unittest_minimal` inside the Jenkinsfile. 

Therefore, I need a macro that does the same thing as `test_step` but does not include the declaration that is part of the `parallel`:  `unittest: CPU MINIMAL: {`

File: tests/scripts/ci.py
Comment (gigiblender) [reply tests/scripts/ci.py:622]: The `_` are replaced with `-` by the `cli_name` function, so the command line flag available and shown in `--help` is `micro-cpp`.

Strangely, if I replace `micro_cpp` with `micro-cpp` then calling, for example, `python tests/scripts/ci.py cpu --micro-cpp` will only run the build. This might be a limitation in how we generate the argument parsers.

File: tests/scripts/task_cpp_unittest.sh
Comment (gigiblender) [reply tests/scripts/task_cpp_unittest.sh]: Update: It seems that there are `crt` tests that get included in the cpp test suite when `USE_MICRO` is enabled. 

Instead of separating those, I am following @areusch advice and writing to a file all the options used to build TVM.

**Review by driazati — APPROVED** [review]
lgtm, waiting for approve from @areusch too before merging

File: tests/python/unittest/test_meta_schedule_task_scheduler.py
Comment (gigiblender) [reply tests/python/unittest/test_meta_schedule_task_scheduler.py:352]: There's this issue open: https://github.com/apache/tvm/issues/12199

Do you mean to add it in a comment?

Comment (driazati) [reply tests/python/unittest/test_meta_schedule_task_scheduler.py:352]: Yeah it should be mentioned in the `reason` so it shows up in the pytest results

File: tests/python/unittest/test_meta_schedule_task_scheduler.py
Comment (zxybazh) [inline tests/python/unittest/test_meta_schedule_task_scheduler.py:352]: Hi, can you please enable this test after rebase to latest main because it’s been fixed, thanks!

**Review by areusch — APPROVED** [review]

**Comment by areusch** [comment]
@gigiblender looks like just need to resolve the merge conflict

**Comment by driazati** [comment]
@tvm-bot merge

**Comment by driazati** [comment]
@tvm-bot merge

**Pull request merged**
